nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -0,0 +1,585 @@
+            .name("Script File")
+            .required(false)
+            .description("Path to script file to execute. Use either file or body not both")
+            .addValidator(new StandardValidators.FileExistsValidator(true))
+            .expressionLanguageSupported(true)
+            .build();
+
+    public static final PropertyDescriptor SCRIPT_BODY = new PropertyDescriptor.Builder()
+            .name("Script Body")
+            .required(false)
+            .description("Body to script to execute. Use either file or body not both")
+            .addValidator(Validator.VALID)
+            .expressionLanguageSupported(true)
+            .build();
+
+    public static final PropertyDescriptor SCRIPT_ARGS = new PropertyDescriptor.Builder()
+            .name("Arguments")
+            .required(false)
+            .description("Arguments to pass to scripting engine")
+            .addValidator(Validator.VALID)
+            .expressionLanguageSupported(true)
+            .defaultValue("")
+            .build();
+
+    public static final PropertyDescriptor MODULES = new PropertyDescriptor.Builder()
+            .name("Module Directory")
+            .description("Path to a directory which contains modules required by the script script.")
+            .required(false)
+            .expressionLanguageSupported(true)
+            .addValidator(new StandardValidators.DirectoryExistsValidator(true, false))
+            .build();
+
+    // A map from engine name to a custom configurator for that engine
+    private final Map<String, ScriptEngineConfigurator> scriptEngineConfiguratorMap = new ConcurrentHashMap<>();
+
+    private final AtomicBoolean isInitialized = new AtomicBoolean(false);
+    private final Lock lock = new ReentrantLock();
+    private SynchronousFileWatcher scriptWatcher;
+
+    private Map<String, ScriptEngineFactory> scriptEngineFactoryMap;
+    private ScriptEngine scriptEngine;
+    private String scriptEngineName;
+    private String scriptPath;
+    private String scriptBody;
+    private String modulePath;
+    private CompiledScript compiledScript;
+    private final AtomicBoolean scriptNeedsReload = new AtomicBoolean(true);
+    private ScheduledExecutorService reloadService;
+    private List<PropertyDescriptor> descriptors;
+
+
+    /**
+     * Initializes this processor. A reload service is defined and scheduled, for the purpose of watching for
+     * script file changes, which indicates a reload is necessary
+     *
+     * @param context in which to perform initialization
+     */
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+    }
+
+
+    protected void createResources() {
+
+        // Set up script file reloader service. This checks to see if the script file has changed, and if so, marks
+        // the script file as needing a reload before evaluation
+        if (reloadService == null) {
+            reloadService = Executors.newScheduledThreadPool(1);
```

Comment:
This Thread Pool is never shutdown. We should have an @OnStopped method that shuts it down & sets it to null so that it can be recreated the next time around - or an @OnShutdown method that terminates it when NiFi is shutdown. May also consider using a custom ThreadFactory that calls Executors.defaultThreadFactory().newThread(r) to create the threads so that you can set the name of the thread so that it is clear from a thread dump what that thread is for.